relayer/relayer.go

```diff
@@ -158,63 +160,61 @@ func NewRelayer(
+	if startBlockHeight == nil {
+		return fmt.Errorf("startBlockHeight cannot be nil")
+	}
+	// Attempt to get the latest processed block height from the database.
+	// Note that the retrieved latest processed block may have already been partially (or fully) processed by the relayer on a previous run. When
+	// processing a warp message in real time, which is when we update the latest processed block in the database, we have no way of knowing
+	// if that is the last warp message in the block
 	latestProcessedBlockData, err := r.db.Get(r.sourceBlockchainID, []byte(database.LatestProcessedBlockKey))
 
-	// The following cases are treated as successful:
+	// First, determine the height to process from. There are two cases:
 	// 1) The database contains the latest processed block data for the chain
-	//    - In this case, we parse the block height and process warp logs from that height to the current block
+	//    - In this case, we process from the maximum of the latest processed block and the configured start block height to the latest block
 	// 2) The database has been configured for the chain, but does not contain the latest processed block data
-	//    - In this case, we save the current block height in the database, but do not process any historical warp logs
+	//    - In this case, we process from the start block height to the latest block
+	height := startBlockHeight
 	if err == nil {
-		// If the database contains the latest processed block data, then back-process all warp messages from that block to the latest block
-		// Note that the retrieved latest processed block may have already been partially (or fully) processed by the relayer on a previous run. When
-		// processing a warp message in real time, which is when we update the latest processed block in the database, we have no way of knowing
-		// if that is the last warp message in the block
+		// Use the max of the latest processed block and the start block height
 		latestProcessedBlock, success := new(big.Int).SetString(string(latestProcessedBlockData), 10)
 		if !success {
 			r.logger.Error("failed to convert latest block to big.Int", zap.Error(err))
 			return err
 		}
+		height = utils.MaxBigInt(latestProcessedBlock, height)
+	} else if !errors.Is(err, database.ErrChainNotFound) && !errors.Is(err, database.ErrKeyNotFound) {
+		// Otherwise, we've encountered an unknown database error
+		r.logger.Warn(
+			"failed to get latest block from database",
+			zap.String("blockchainID", r.sourceBlockchainID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
 
-		err = sub.ProcessFromHeight(latestProcessedBlock)
-		if err != nil {
-			r.logger.Warn(
-				"Encountered an error when processing historical blocks. Continuing to normal relaying operation.",
-				zap.String("blockchainID", r.sourceBlockchainID.String()),
-				zap.Error(err),
-			)
-		}
-		return nil
+	// Process from the determined height to the latest block.
+	if height.Cmp(big.NewInt(0)) == 0 {
```

Comment:
If teh last processed block height is not in db, should we require that the config starting block height is set, instead of automatically starting from 0 and sending a warning?

Reply:
See the discussion here: https://github.com/ava-labs/awm-relayer/pull/114#discussion_r1443050931

Reply:
yup I agree with a majority on that approach, but didn't see the reason for starting from 0 if both are not set, probably was discussed offline.

Reply:
Thinking about this again, your suggestion might be a better UI. When starting a new relayer for the first time, I think it's more likely that a missing `StartBlockHeight` value is user error rather than intentionally starting processing blocks from genesis. If the user does wish to process from genesis, they can always set `StartBlockHeight` to 0 explicitly. @michaelkaplan13 curious what your thoughts are in case I'm missing something, otherwise I'm going to take @minghinmatthewlam's suggestion.

Reply:
Implemented this in a separate PR for ease of review: https://github.com/ava-labs/awm-relayer/pull/133

Reply:
Yep I agree with @minghinmatthewlam's suggestion here.